src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,157 @@
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            var getFixesForSpanTasks = diagnostics
+                .GroupBy(diagnostic => diagnostic.Location.SourceSpan)
+                .Select(diagnosticsForSourceSpan => addImportService
+                    .GetFixesForDiagnosticsAsync(document, diagnosticsForSourceSpan.Key, diagnosticsForSourceSpan.AsImmutable(), symbolSearchService, true, packageSources, cancellationToken)
+                );
+
+            await Task.WhenAll(getFixesForSpanTasks).ConfigureAwait(false);
+
+            var fixes = ArrayBuilder<AddImportFixData>.GetInstance();
+            foreach (var getFixesForSpanTask in getFixesForSpanTasks)
+            {
+                var fixesForSpan = await getFixesForSpanTask.ConfigureAwait(false);
+
+                foreach (var fixForSpan in fixesForSpan)
+                {
+                    // If there is more than one fix then we will leave it for the user to manually apply
+                    if (fixForSpan.Fixes.Length == 1)
+                    {
+                        fixes.Add(fixForSpan.Fixes[0]);
+                    }
+                }
+            }
+
+            return fixes.ToImmutableAndFree();
+        }
+
+        private async Task<Document> ApplyFixesAsync(Document document, ImmutableArray<AddImportFixData> fixes, CancellationToken cancellationToken)
+        {
+            if (fixes.IsEmpty
+                || !document.TryGetText(out var text))
+            {
+                return document;
+            }
+
+            var solution = document.Project.Solution;
+            var packageInstallerService = solution.Workspace.Services.GetService<IPackageInstallerService>();
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            // Do not limit the results since we plan to fix all the reported issues.
+            var codeActions = addImportService.GetCodeActionsForFixes(document, fixes, packageInstallerService, limitResults: false);
+            var getChangesTasks = codeActions.Select(action => GetChangesForCodeActionAsync(document, action, cancellationToken));
+
+            await Task.WhenAll(getChangesTasks).ConfigureAwait(false);
+
+            var allTextChanges = Enumerable.Empty<TextChange>();
+            var allAddedProjectReferences = Enumerable.Empty<ProjectReference>();
+            var allAddedMetaDataReferences = Enumerable.Empty<MetadataReference>();
+
+            foreach (var getChangesTask in getChangesTasks)
+            {
+                var (projectChanges, textChanges) = await getChangesTask.ConfigureAwait(false);
+
+                allTextChanges = allTextChanges.Concat(textChanges);
+                allAddedProjectReferences = allAddedProjectReferences.Concat(projectChanges.GetAddedProjectReferences());
+                allAddedMetaDataReferences = allAddedMetaDataReferences.Concat(projectChanges.GetAddedMetadataReferences());
```

Comment:
we should use sets for these so we only have references one time.   A lot simpler than concats of contacts with distincts.